ipsec_offload_plugin/ipsec_offload/ipsec_offload.c

```diff
@@ -801,31 +805,53 @@ METHOD(kernel_ipsec_t, add_policy, status_t,
+						       data->sa->mode == MODE_TUNNEL, true);
 		if(err == IPSEC_FAILURE) {
 			DBG2(DBG_KNL, "Inline_crypto_ipsec ipsec_tx_sa_classification_table:"
-			     "add entry failed");
+			     "add entry failed for underlay");
 		} else if (err == IPSEC_DUP_ENTRY) {
 			err = ipsec_tx_sa_classification_table(IPSEC_TABLE_MOD,
 							       dst, src, 1,
 							       offload_id, offload_id,
 							       id->src_ts->get_protocol(id->src_ts),
-							       data->sa->mode == MODE_TUNNEL);
+							       data->sa->mode == MODE_TUNNEL, true);
 			if(err == IPSEC_FAILURE)
 				DBG2(DBG_KNL, "Inline_crypto_ipsec ipsec_tx_sa_classification_table:"
-				     "Modify entry failed");
+				     "Modify entry failed for underlay");
 
 			if (!reqid_bitset(data->sa->reqid))
 				DBG1(DBG_KNL, "ipsec_tx_sa_classification_table: Failed to set the reqid bit!!");
 		} else
-			DBG2(DBG_KNL, "ipsec_tx_sa_classification_table add entry done");
+			DBG2(DBG_KNL, "ipsec_tx_sa_classification_table add entry done for underlay");
+
+                err = ipsec_tx_sa_classification_table(IPSEC_TABLE_ADD,
+                                                       dst, src, 1,
+                                                       offload_id, offload_id,
+                                                       id->src_ts->get_protocol(id->src_ts),
+                                                       data->sa->mode == MODE_TUNNEL, false);
+                if(err == IPSEC_FAILURE) {
+                        DBG2(DBG_KNL, "Inline_crypto_ipsec ipsec_tx_sa_classification_table:"
+                             "add entry failed for non underlay");
+                } else if (err == IPSEC_DUP_ENTRY) {
+                        err = ipsec_tx_sa_classification_table(IPSEC_TABLE_MOD,
+                                                               dst, src, 1,
+                                                               offload_id, offload_id,
+                                                               id->src_ts->get_protocol(id->src_ts),
+                                                               data->sa->mode == MODE_TUNNEL, false);
+                        if(err == IPSEC_FAILURE)
+                                DBG2(DBG_KNL, "Inline_crypto_ipsec ipsec_tx_sa_classification_table:"
+                                     "Modify entry failed for non underlay");
 
+                        if (!reqid_bitset(data->sa->reqid))
+                                DBG1(DBG_KNL, "ipsec_tx_sa_classification_table: Failed to set the reqid bit!!");
+                } else
+                        DBG2(DBG_KNL, "ipsec_tx_sa_classification_table add entry done for non underlay");
 
 		if (data->sa->mode == MODE_TUNNEL) {
 			err = ipsec_outer_ipv4_encap_mod_table(IPSEC_TABLE_ADD,
 							       offload_id,
 							       src_outer, dst_outer,
-							       id->src_ts->get_protocol(id->src_ts),
+							       PROTO_IP, // proto should be 0x04 in tunnel mode for encap_mod_table
```

Comment:
Don't hardcode this value. We should get that value from "id" or data.

Reply:
Addressed